README.md
habrok [![Build Status](https://travis-ci.org/kurttheviking/habrok-js.svg?branch=master)](https://travis-ci.org/kurttheviking/habrok-js) [![Coverage Status](https://coveralls.io/repos/github/kurttheviking/habrok-js/badge.svg?branch=master)](https://coveralls.io/github/kurttheviking/habrok-js?branch=master)
======

[Promises/A+](https://promisesaplus.com/), [request](https://www.npmjs.com/package/request)-powered, [boom](https://www.npmjs.com/package/boom)-enabled, JSON-first HTTP API client with automatic retry


## Quick start

### Prerequisites

- [Node.js LTS (8+)](https://nodejs.org/)
- [npm](https://www.npmjs.com/)

### Installing

Install with `npm`:

```sh
npm install habrok --save
```

### Example

```js
const Habrok = require('habrok');

const habrok = new Habrok();

habrok.request({
  method: 'GET',
  uri: 'https://api.github.com/repositories'
})
.then(console.log);
```

```json
{
  "statusCode": 200,
  "headers": {
    "content-length": "4477",
    "content-type": "application/json; charset=utf-8",
    "etag": "6ffc6a0dbbe2613e4d8b3f7444e5c604",
    ...
  },
  "body": [
    {
      "id": 1296269,
      "private": false,
      "owner": {
        "id": 1,
        "login": "octocat",
        "avatar_url": "https://github.com/images/error/octocat_happy.gif",
        ...
      },
      "name": "hello-World",
      "full_name": "octocat/hello-world",
      "description": "Octocat's first repo!",
      "url": "https://api.github.com/repos/octocat/hello-world",
      ...
    },
    ...
  ]
}
```


## API

### `Habrok` (Constructor)

#### Definition

```js
Habrok([configuration])
```

#### Required arguments

*None*

#### Optional arguments

- `configuration`: `Object` with one or more of the below properties:

| Property | Type | Description | Default |
| :------- | :--- | :---------- | :------ |
| `disableAutomaticJson` | `Boolean` | Disable JSON headers and request/response bodies | `false` |
| `disableCustomHeaders` | `Boolean` | Disable request headers added by Habrok | `false` |
| `disableRetryEconnreset` | `Boolean` | Disable retry for connection reset errors | `false` |
| `retries` | `Number` | Number of times to retry a failed request | `5` |
| `retryMinDelay` | `Number` | Minimum milliseconds to wait before retrying a request | `100` |
| `retryMaxDelay` | `Number` | Maximum milliseconds to wait before retrying a request | *None* |
| `retryCodes` | `Array<Number>` | 	HTTP status codes that trigger a retry | `[429, 502, 503, 504]` |

By default, each request includes the following headers (which can be prevented with `disableCustomHeaders`):

- `User-Agent`: `habrok/[MAJOR.MINOR.PATCH]`, versioned according to [`package.json`](./package.json) (e.g. `habrok/1.0.0`)
- `X-Node-Platform`: Directly from [`process.platform`](https://nodejs.org/api/process.html#process_process_platform) (e.g. `linux`)
- `X-Node-Version`: Derived from [`process.version`](https://nodejs.org/api/process.html#process_process_version) (e.g. `6.11.1`)

The retry logic follows exponential backoff, summarized as:

```
MINIMUM(retryMaxDelay, retryMinDelay * (attempt ** 2))
```

With the default `configuration`, a failing request observes a delay sequence of `100`, `400`, `900`, and `1600` milliseconds before rejecting with an error.

#### Returns

A Habrok HTTP API Client instance.

#### Examples

Construct a default client:

```js
const Habrok = require('habrok');

const habrok = new Habrok();
```

Construct a client with a minimum retry delay of 250 milliseconds:

```js
const Habrok = require('habrok');

const habrok = new Habrok({ retryMinDelay: 250 });
```

Construct a client that does not send Habrok-generated headers:

```js
const Habrok = require('habrok');

const habrok = new Habrok({ disableCustomHeaders: true });
```

### `Habrok#request`

#### Definition

```js
habrok.request(req[, options])
```

#### Required arguments

- `req`: `Object`, a [request-compatible](https://www.npmjs.com/package/request#requestoptions-callback) object

#### Optional arguments

- `options`: `Object` with one or more of the below properties:

| Property | Type | Description | Default |
| :------- | :--- | :---------- | :------ |
| `attempt` | `Number` | Integer indicating current request sequence number | *None* |

Generally, `attempt` is not needed. The internal retry engine will pass the current attempt count into the next request. Override only as necessary &ndash; e.g. in cases where the retry logic should be bypassed.

#### Returns

A `Promise` that resolves to an `Object` with the following properties:

- `statusCode`: `Number`, the HTTP status code provided in the response
- `headers`: `Object`, HTTP headers (lower-cased) and their values provided in the response
- `body`: `Any`, the JSON-parsed response body (or the raw body if `disableAutomaticJson` was set)

The `Promise` is rejected with a [Boom](https://www.npmjs.com/package/boom)-wrapped error if an HTTP error occurs. The `Promise` is rejected with a generic `Error` if an error is returned by the underlying request library (usually from [http.ClientRequest](https://nodejs.org/api/http.html#http_class_http_clientrequest)).

#### Examples

Send a `GET` request:

```js
habrok.request({
  method: 'GET',
  uri: 'https://api.viki.ng/longships'
})
```

Send a `POST` request:

```js
habrok.request({
  method: 'POST',
  uri: 'https://api.viki.ng/longships',
  json: {
    name: 'Oseberg'
  }
})
```


## Development

### Debug

The [debug](https://www.npmjs.com/package/debug) module is used for runtime logging. Omit the `DEBUG` environment variable to squelch all logging. Set `DEBUG` to the desired level (e.g. `DEBUG=habrok`) to restrict logging to a desired service. Or, use `DEBUG=*` to get all debug output from everywhere, including dependencies.

```sh
DEBUG=habrok* node index
```

### Tests

To run the unit tests:

```sh
npm test
```

This project maintains ~100% coverage of statements, branches, and functions. To determine unit test coverage:

```sh
npm run coverage
```

### Contribute

PRs are welcome! PRs must pass unit tests and linting prior to merge. PRs must not reduce unit coverage. For bugs, please include a failing test which passes when your PR is applied. To enable a git hook that runs `npm test` prior to pushing, `cd` into your repo and run:

```sh
touch .git/hooks/pre-push
chmod +x .git/hooks/pre-push
echo "npm test" > .git/hooks/pre-push
```

### Versioning

This project follows [semantic versioning](http://semver.org/). See the [changelog](CHANGELOG.md) for release information.


## License

- This module is licensed under the [ISC License](./LICENSE)
- The underlying `boom` module is licensed under the [BSD 3-Clause License](https://github.com/hapijs/boom/blob/8b35a4c5f6dc706f3396cfed3f5cc5f60a5f6eb1/LICENSE)
- The underlying `debug` module is licensed under the [MIT License](https://github.com/visionmedia/debug/blob/a45d4a0239f071634804dd7901dd33b0d0d407c9/LICENSE)
- The underlying `has` module is licensed under the [MIT License](https://github.com/tarruda/has/blob/535c5c8ed1dc255c9e223829e702548dd514d2a5/LICENSE-MIT)
- The underlying `request` module is licensed under the [Apache 2.0](https://github.com/request/request/blob/643c43ab7be269a1efaa080ff05a18fff3f64cd7/LICENSE)


## Etymology

From [Grímnismál](https://en.wikipedia.org/wiki/Gr%C3%ADmnism%C3%A1l), [Stanza 44](http://www.voluspa.org/grimnismal41-45.htm):

> The best of trees must Yggdrasil be,
> Skithblathnir best of boats;
> Of all the gods is Odin the greatest,
> And Sleipnir the best of steeds;
> Bifrost of bridges, Bragi of skalds,
> Habrok of hawks, and Garm of hounds.

Habrok, data-in-flight at its best.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.


## 4.1.0 &ndash; July 28, 2019

- :arrow_up: Upgrade [boom](https://www.npmjs.com/package/@hapi/boom)


## 4.0.0 &ndash; July 28, 2019

- :arrow_up: Upgrade dependencies
- :boom: Change minimum support target to Node 8 (LTS maintenance)


## 3.0.0 &ndash; January 26, 2019

- :arrow_up: Upgrade dependencies
- :boom: Change support target to match current maintenance LTS (Node 6)


## 2.1.0 &ndash; September 11, 2017

- :sparkles: Add support for automatic retry of [`ECONNRESET` errors](https://github.com/nodejs/node/blob/640b20616d2cdc46bc3df8703cdc1395578ff1b3/lib/_http_client.js#L343)
- :arrow_up: Upgrade dependencies


## 2.0.2 &ndash; July 26, 2017

- :hammer: Improve readability of response `debug` output of


## 2.0.1 &ndash; July 26, 2017

- :books: Clarify CHANGELOG and README examples


## 2.0.0 &ndash; July 26, 2017

- :boom: `Habrok#request` resolves to an `Object` with `statusCode`, `headers`, and `body` properties


## 1.0.0 &ndash; July 25, 2017

- :tada: Initial release


test/lint.test.js
/* eslint-disable global-require, strict */

require('eslint');  // [KE] establish the dependency for dependency checkers

const lint = require('mocha-eslint');

const options = {
  formatter: 'compact'
};

const paths = [
  'index.js',
  'lib'
];

lint(paths, options);


test/Habrok.test.js
/* global describe, it */
/* eslint-disable global-require, strict */

const expect = require('chai').expect;

describe('Habrok', () => {
  it('is a constructor', () => {
    const Habrok = require('../index');

    expect(Habrok).to.be.a('function');
  });

  describe('default configuration', () => {
    const Habrok = require('../index');
    const habrok = new Habrok();

    it('sets the DISABLE_AUTO_JSON constant', () => {
      expect(habrok.DISABLE_AUTO_JSON).to.equal(false);
    });

    it('sets the DISABLE_CUSTOM_HEADERS constant', () => {
      expect(habrok.DISABLE_CUSTOM_HEADERS).to.equal(false);
    });

    it('sets the DISABLE_CUSTOM_HEADERS constant', () => {
      expect(habrok.DISABLE_CUSTOM_HEADERS).to.equal(false);
    });

    it('sets the DISABLE_RETRY_ECONNRESET constant', () => {
      expect(habrok.DISABLE_RETRY_ECONNRESET).to.equal(false);
    });

    it('sets the RETRIES constant', () => {
      expect(habrok.RETRIES).to.equal(5);
    });

    it('sets the RETRY_CODES constant', () => {
      expect(habrok.RETRY_CODES).to.deep.equal([429, 502, 503, 504]);
    });

    it('sets the RETRY_MS_MAX constant', () => {
      expect(habrok.RETRY_MS_MAX).to.equal(null);
    });

    it('sets the RETRY_MS_MIN constant', () => {
      expect(habrok.RETRY_MS_MIN).to.equal(100);
    });

    it('has a request method', () => {
      expect(habrok.request).to.be.a('function');
    });
  });

  describe('explicit configuration', () => {
    const Habrok = require('../index');

    const config = {
      disableAutomaticJson: true,
      disableCustomHeaders: true,
      disableRetryEconnreset: true,
      retries: 3,
      retryCodes: [404, '503'],
      retryMaxDelay: 1 + parseInt(Math.random() * 1000, 10),
      retryMinDelay: 1 + parseInt(Math.random() * 1000, 10)
    };

    const habrok = new Habrok(config);

    it('sets the DISABLE_AUTO_JSON constant', () => {
      expect(habrok.DISABLE_JSON_BODY).to.equal(config.disableJsonBody);
    });

    it('sets the DISABLE_CUSTOM_HEADERS constant', () => {
      expect(habrok.DISABLE_CUSTOM_HEADERS).to.equal(config.disableCustomHeaders);
    });

    it('sets the DISABLE_RETRY_ECONNRESET constant', () => {
      expect(habrok.DISABLE_RETRY_ECONNRESET).to.equal(config.disableRetryEconnreset);
    });

    it('sets the RETRIES constant', () => {
      expect(habrok.RETRIES).to.equal(config.retries);
    });

    it('sets the RETRY_CODES constant', () => {
      expect(habrok.RETRY_CODES).to.deep.equal([404, 503]);
    });

    it('sets the RETRY_MS_MAX constant', () => {
      expect(habrok.RETRY_MS_MAX).to.equal(config.retryMaxDelay);
    });

    it('sets the RETRY_MS_MIN constant', () => {
      expect(habrok.RETRY_MS_MIN).to.equal(config.retryMinDelay);
    });

    it('has a request method', () => {
      expect(habrok.request).to.be.a('function');
    });
  });
});


test/habrok/request.test.js
/* global describe, it, beforeEach, afterEach */
/* eslint-disable global-require, strict */

'use strict';

const { expect } = require('chai');
const mockery = require('mockery');
const sinon = require('sinon');
const uuid = require('uuid-with-v6');

describe('Habrok#request', () => {
  const body = { x: uuid.v4() };
  const res = { statusCode: 200, headers: { z: uuid.v4() } };

  let habrok;
  let request;

  beforeEach(() => {
    mockery.enable({
      warnOnReplace: false,
      warnOnUnregistered: false,
      useCleanCache: true
    });

    request = sinon.stub().yields(null, res, body);
    mockery.registerMock('request', request);

    habrok = require('../../index')();
  });

  afterEach(() => {
    mockery.deregisterAll();
    mockery.disable();
  });

  it('invokes requestjs with a valid method and uri', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri }).then(() => {
      const req = request.getCall(0).args[0];

      expect(req.method).to.equal(method);
      expect(req.uri).to.equal(uri);
    });
  });

  it('invokes requestjs with default headers', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri }).then(() => {
      const req = request.getCall(0).args[0];

      expect(req.headers['User-Agent']).to.match(/^habrok/);
      expect(req.headers).to.have.property('X-Node-Platform');
      expect(req.headers).to.have.property('X-Node-Version');
    });
  });

  it('invokes requestjs with default json parameter', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri }).then(() => {
      const req = request.getCall(0).args[0];

      expect(req.json).to.equal(true);
    });
  });

  it('resolves to the expected response format', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri }).then((out) => {
      expect(out).to.deep.equal(Object.assign({}, res, { body }));
    });
  });
});

describe('Habrok#request with disabled custom headers', () => {
  const body = { x: uuid.v4() };
  const res = { statusCode: 200, headers: { z: uuid.v4() } };

  let habrok;
  let request;

  beforeEach(() => {
    mockery.enable({
      warnOnReplace: false,
      warnOnUnregistered: false,
      useCleanCache: true
    });

    request = sinon.stub().yields(null, res, body);
    mockery.registerMock('request', request);

    habrok = require('../../index')({ disableCustomHeaders: true });
  });

  afterEach(() => {
    mockery.deregisterAll();
    mockery.disable();
  });

  it('invokes requestjs without default headers', () => {
    const headers = { z: uuid.v4() };
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ headers, method, uri }).then(() => {
      const req = request.getCall(0).args[0];

      expect(req.headers).to.not.have.property('User-Agent');
      expect(req.headers).to.not.have.property('X-Node-Platform');
      expect(req.headers).to.not.have.property('X-Node-Version');
    });
  });

  it('invokes requestjs with invocation headers', () => {
    const headers = { z: uuid.v4() };
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ headers, method, uri }).then(() => {
      const req = request.getCall(0).args[0];

      expect(req.headers).to.deep.equal(headers);
    });
  });
});

describe('Habrok#request with disabled JSON parsing', () => {
  const body = `<x>${uuid.v4()}</x>`;
  const res = { statusCode: 200, headers: { z: uuid.v4() } };

  let habrok;
  let request;

  beforeEach(() => {
    mockery.enable({
      warnOnReplace: false,
      warnOnUnregistered: false,
      useCleanCache: true
    });

    request = sinon.stub().yields(null, res, body);
    mockery.registerMock('request', request);

    habrok = require('../../index')({ disableAutomaticJson: true });
  });

  afterEach(() => {
    mockery.deregisterAll();
    mockery.disable();
  });

  it('invokes requestjs without json parameter', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri }).then(() => {
      const req = request.getCall(0).args[0];

      expect(req.json).to.equal(undefined);
    });
  });

  it('resolves to the expected response format', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri }).then((out) => {
      expect(out).to.deep.equal(Object.assign({}, res, { body }));
    });
  });
});

describe('Habrok#request with a retried HTTP error (429)', () => {
  const body = { x: uuid.v4() };
  const res = { statusCode: 429, headers: { z: uuid.v4() } };

  let habrok;
  let request;

  beforeEach(() => {
    mockery.enable({
      warnOnReplace: false,
      warnOnUnregistered: false,
      useCleanCache: true
    });

    request = sinon.stub().yields(null, res, body);
    mockery.registerMock('request', request);

    habrok = require('../../index')({ retryMinDelay: 0 });
  });

  afterEach(() => {
    mockery.deregisterAll();
    mockery.disable();
  });

  it('retries request up to retry limit', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch(() => {
        expect(request.callCount).to.equal(habrok.RETRIES);
      });
  });

  it('rejects with a Boom#tooManyRequests error', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch((err) => {
        expect(err).to.match(/Too Many Requests/);
        expect(err.isBoom).to.equal(true);
        expect(err.output.statusCode).to.equal(429);
      });
  });

  it('rejected Boom#tooManyRequests contains response body', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch((err) => {
        expect(err.data).to.deep.equal(body);
      });
  });

  it('rejects within expected elapsed time', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    const start = Date.now();

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch(() => {
        // [KE] rough heuristic; empirically the observed elapsed time is between 9-10ms;
        //      this test is sufficient to demonstrate elapsed time is substantially below default
        const observed = Date.now() - start;

        expect(observed).to.be.below(20);
      });
  });
});

describe('Habrok#request with a retried HTTP error (429) and max wait', () => {
  const body = { x: uuid.v4() };
  const res = { statusCode: 429, headers: { z: uuid.v4() } };

  let habrok;
  let request;

  beforeEach(() => {
    mockery.enable({
      warnOnReplace: false,
      warnOnUnregistered: false,
      useCleanCache: true
    });

    request = sinon.stub().yields(null, res, body);
    mockery.registerMock('request', request);

    habrok = require('../../index')({ retryMaxDelay: 0 });
  });

  afterEach(() => {
    mockery.deregisterAll();
    mockery.disable();
  });

  it('retries request up to retry limit', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch(() => {
        expect(request.callCount).to.equal(habrok.RETRIES);
      });
  });

  it('rejects within expected elapsed time', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    const start = Date.now();

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch(() => {
        // [KE] rough heuristic; empirically the observed elapsed time is between 9-10ms;
        //      this test is sufficient to demonstrate elapsed time is substantially below default
        const observed = Date.now() - start;

        expect(observed).to.be.below(20);
      });
  });
});

describe('Habrok#request with a non-retried HTTP error (500)', () => {
  const body = { x: uuid.v4() };
  const res = { statusCode: 500, headers: { z: uuid.v4() } };

  let habrok;
  let request;

  beforeEach(() => {
    mockery.enable({
      warnOnReplace: false,
      warnOnUnregistered: false,
      useCleanCache: true
    });

    request = sinon.stub().yields(null, res, body);
    mockery.registerMock('request', request);

    habrok = require('../../index')({ retryMinDelay: 1 });
  });

  afterEach(() => {
    mockery.deregisterAll();
    mockery.disable();
  });

  it('does not retry request', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch(() => {
        expect(request.callCount).to.equal(1);
      });
  });

  it('rejects with a Boom#badImplementation error', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch((err) => {
        expect(err).to.match(/Internal Server Error/);
        expect(err.isBoom).to.equal(true);
        expect(err.output.statusCode).to.equal(500);
      });
  });

  it('rejected Boom#badImplementation contains response body', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch((err) => {
        expect(err.data).to.deep.equal(body);
      });
  });
});

describe('Habrok#request with a retried ECONNRESET error', () => {
  const body = { x: uuid.v4() };
  const requestError = new Error('ECONNRESET');
  const res = {};

  requestError.code = 'ECONNRESET';

  let habrok;
  let request;

  beforeEach(() => {
    mockery.enable({
      warnOnReplace: false,
      warnOnUnregistered: false,
      useCleanCache: true
    });

    request = sinon.stub().yields(requestError, res, body);
    mockery.registerMock('request', request);

    habrok = require('../../index')({ retryMinDelay: 0 });
  });

  afterEach(() => {
    mockery.deregisterAll();
    mockery.disable();
  });

  it('retries request up to retry limit', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch(() => {
        expect(request.callCount).to.equal(habrok.RETRIES);
      });
  });

  it('rejects with the request error', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch((err) => {
        expect(err).to.match(/ECONNRESET/);
        expect(err.isBoom).to.equal(undefined);
      });
  });

  it('rejects within expected elapsed time', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    const start = Date.now();

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch(() => {
        // [KE] rough heuristic; empirically the observed elapsed time is between 9-10ms;
        //      this test is sufficient to demonstrate elapsed time is substantially below default
        const observed = Date.now() - start;

        expect(observed).to.be.below(20);
      });
  });
});

describe('Habrok#request with request.js error', () => {
  let habrok;
  let request;

  beforeEach(() => {
    mockery.enable({
      warnOnReplace: false,
      warnOnUnregistered: false,
      useCleanCache: true
    });

    request = sinon.stub().yields(new Error('invalid input'), {}, {});
    mockery.registerMock('request', request);

    habrok = require('../../index')();
  });

  afterEach(() => {
    mockery.deregisterAll();
    mockery.disable();
  });

  it('does not retry request', () => {
    const method = undefined;
    const uri = undefined;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch(() => {
        expect(request.callCount).to.equal(1);
      });
  });

  it('rejects with a generic error', () => {
    const method = 'GET';
    const uri = `https://api.viki.ng/longships/${uuid.v4()}`;

    return habrok.request({ method, uri })
      .then(() => { throw new Error('fail test'); })
      .catch((err) => {
        expect(err).to.match(/invalid input/);
        expect(err.isBoom).to.equal(undefined);
      });
  });
});


lib/Habrok.js
const cwd = __dirname;

const Boom = require('boom');
const debug = require('debug')('habrok');
const has = require('has');
const path = require('path');
const requestjs = require('request');

const pjson = require(path.join(cwd, '../package.json')); // eslint-disable-line import/no-dynamic-require, max-len

const integer = require('./utils/castToInteger');

const ECONNRESET = 'ECONNRESET';
const USER_AGENT = `habrok/${pjson.version}`;
const NODE_VERSION = process.version.substr(1);
const NODE_PLATFORM = process.platform;

function Habrok(config) {
  if (!(this instanceof Habrok)) {
    return new Habrok(config);
  }

  debug('initialize', config);

  const conf = config || {};

  this.DISABLE_AUTO_JSON = Boolean(conf.disableAutomaticJson);
  this.DISABLE_CUSTOM_HEADERS = Boolean(conf.disableCustomHeaders);
  this.DISABLE_RETRY_ECONNRESET = Boolean(conf.disableRetryEconnreset);
  this.RETRIES = has(conf, 'retries') ? integer(conf.retries, { min: 0 }) : 5;
  this.RETRY_CODES = has(conf, 'retryCodes') ? [].concat(conf.retryCodes).map(integer) : [429, 502, 503, 504]; // eslint-disable-line max-len
  this.RETRY_MS_MAX = has(conf, 'retryMaxDelay') ? integer(conf.retryMaxDelay, { min: 0 }) : null;
  this.RETRY_MS_MIN = has(conf, 'retryMinDelay') ? integer(conf.retryMinDelay, { min: 0 }) : 100;

  return this;
}

Habrok.prototype.request = function request(raw, opts) {
  debug('prepare', raw, opts);

  const options = opts || {};

  const attempt = has(options, 'attempt') ? integer(options.attempt, { min: 0 }) : 1;
  const req = Object.assign({}, raw);
  const self = this;

  if (self.DISABLE_CUSTOM_HEADERS === false) {
    req.headers = Object.assign(
      {
        'User-Agent': USER_AGENT,
        'X-Node-Platform': NODE_PLATFORM,
        'X-Node-Version': NODE_VERSION
      },
      req.headers
    );
  }

  if (self.DISABLE_AUTO_JSON === false && has(req, 'json') === false) {
    req.json = true;
  }

  debug('request', req);

  return new Promise((resolve, reject) => requestjs(req, (err, res, body) => {
    // [KE] if/else is purely for debug so ignore code coverage requirements
    /* istanbul ignore next */
    if (res) {
      debug('response', req.method, req.uri || req.url, res.statusCode, body);
    } else {
      debug('no response');
    }

    const errorCode = (err && err.code) || null;
    const statusCode = (res && res.statusCode) || 500;

    if (err && (errorCode !== ECONNRESET || self.disableRetryEconnreset)) {
      return reject(err);
    }

    if (errorCode === ECONNRESET || statusCode > 299) {
      if ((!err && self.RETRY_CODES.indexOf(statusCode) === -1) || attempt >= self.RETRIES) {
        debug('fail', req);

        return reject(err || new Boom(null, { statusCode, data: body }));
      }

      return setTimeout(
        () => {
          debug('retry', raw);
          resolve(self.request(raw, Object.assign({}, options, { attempt: attempt + 1 })));
        },
        integer(
          self.RETRY_MS_MIN * Math.pow(attempt, 2), // eslint-disable-line no-restricted-properties
          { max: self.RETRY_MS_MAX }
        )
      );
    }

    debug('succees', req);

    return resolve({
      statusCode: res.statusCode,
      headers: res.headers,
      body
    });
  }));
};

module.exports = Habrok;
